include/revng/PTML/CommentPlacementHelper.h

```diff
@@ -0,0 +1,354 @@
+
+template<typename NodeType>
+  requires yield::HasStatementTraits<NodeType *>
+struct GraphTraits<llvm::DomTreeNodeBase<NodeType> *>
+  : public DomTreeGraphTraitsBase<
+      llvm::DomTreeNodeBase<NodeType>,
+      typename llvm::DomTreeNodeBase<NodeType>::const_iterator> {};
+
+template<typename NodeType>
+  requires yield::HasStatementTraits<NodeType *>
+struct GraphTraits<const llvm::DomTreeNodeBase<NodeType> *>
+  : public DomTreeGraphTraitsBase<
+      const llvm::DomTreeNodeBase<NodeType>,
+      typename llvm::DomTreeNodeBase<NodeType>::const_iterator> {};
+
+} // namespace llvm
+
+namespace yield {
+
+/// This is a helper for deciding which statement (within a given graph) is
+/// best suited for emission of a given comment.
+///
+/// \tparam NodeType the type of the graph node.
+///         Note that it has to specialize `StatementTraits`.
+///
+/// A typical usage would be something like:
+///
+/// ```cpp
+///   // 1. Make a helper object from your function and your graph:
+///   CommentPlacementHelper<MyNodeType> CommentPlacement(MyModelFunction,
+///                                                       MyGraph);
+///   // note: `MyNodeType` must either have a `getParent()` method returning
+///   //       `decltype(MyGraph) *` OR provide a specialization of
+///   //       `llvm::DomTreeNodeTraits`.
+///
+///   // 2. When emitting your artifact, add calls to `getComments`:
+///   for (const auto &Statement : myWayOfIteratingTheGraph(MyGraph)) {
+///     // . . .
+///     for (const auto &Comment : CommentPlacement.getComments(Statement)) {
+///       emitTheComment(Comment);
+///       // note: see `PTML/Doxygen` for the standard way of emitting comments
+///     }
+///     // . . .
+///   }
+///
+///   // 3. Don't forget to emit homeless comments somewhere.
+///   for (const auto &Comment : CommentPlacement.getHomelessComments())
+///     emitTheComment(Comment);
+/// ```
+///
+/// With the interface out of the way, lets now discuss _how_ statements are
+/// selected:
+///
+/// 1. For each comment-statement pair a similarity score is computed
+///    (see \ref tverskyIndex) by traversing the graph.
+/// 2. Then, for each comment a single _address set_ (note: not a statement yet)
+///    is selected.
+/// 3. Then graph is traversed again, this time looking for all the statements
+///    whose location exactly matches the selected one, except for statements
+///    that are reachable in a dominator tree from an already selected
+///    statement.
+/// 4. All these statements are recorded in a \ref ResultMap and will be emitted
+///    once requested by a \ref getComments call.
+template<HasStatementTraits NodeType>
+class CommentPlacementHelper {
+private:
+  using Trait = StatementTraits<NodeType>;
+  using InternalNodeType = std::remove_pointer_t<NodeType>;
+
+  using StatementComment = model::StatementComment;
+  using CLT = decltype(std::declval<StatementComment>().Location());
+  using CommentLocationType = std::decay_t<CLT>;
```

Comment:
Let's just have a `using CommentLocation = ...` somewhere, no need for this much abstraction. Makes things less obvious to the newcomer without much benefit.

Reply:
Actually we already do: `model::StatementComment::TypeOfLocation`.

My bad on forgetting it exists.